pkg/events/step/reconciler.go

```diff
@@ -0,0 +1,135 @@
+package step
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	v1beta1client "github.com/tektoncd/pipeline/pkg/client/clientset/versioned/typed/pipeline/v1beta1"
+
+	"github.com/tom24d/step-observe-controller/pkg/events/step/resources"
+)
+
+const (
+	annotationKey = "tom24d.plugin.step-observer/result"
+)
+
+// ReconcileStepEvent is entry point to reconcile taskrun to determine whether it should emit CloudEvent.
+func ReconcileStepEvent(ctx context.Context, taskrun *v1beta1.TaskRun, client v1beta1client.TaskRunInterface) error {
+	logger := logging.FromContext(ctx)
+	if taskrun.Status.TaskSpec == nil ||
+		len(GetSteps(taskrun)) < 1 {
+		logger.Infof("step events emission skipped as no step in taskrun: %s", taskrun.Name)
+		return nil
+	}
+
+	sent, err := initializeAnnotation(taskrun)
+	if err != nil {
+		logger.Fatalf("failed to initialize annotation: %v", err)
+		return nil
+	}
+	logger.Infof("RECEIVED annotation: %s, annotation in the run: %v", sent, taskrun.Annotations)
+
+
+	for i, step := range GetStepStatuses(taskrun) {
+		// start
+		if step.ContainerState.Running != nil {
+			// (Running) emit&mark started if i=0
+			if i == 0 && !taskrun.Status.StartTime.IsZero(){
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepStarted)
+			}
+		}
+		// emit&mark started if i!=0 && i-1 step marked as succeeded
+		if i != 0 && sent.IsMarked(GetStepStatuses(taskrun)[i-1].Name, resources.CloudEventTypeStepSucceeded) {
+			ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepStarted)
+		}
+
+		if step.ContainerState.Terminated != nil {
+			// skipped
+			// (Terminated) emit&mark skipped if i!=0 && i-1 step marked as failed||skipped,
+			// then continue to avoid being considered as failure
+			if i != 0 &&
+					(sent.IsMarked(GetStepStatuses(taskrun)[i-1].Name, resources.CloudEventTypeStepFailed) ||
+					sent.IsMarked(GetStepStatuses(taskrun)[i-1].Name, resources.CloudEventTypeStepSkipped)) {
+
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepSkipped)
+				continue
+			}
+			// succeeded
+			// (Terminated) emit&mark succeeded if exit code is 0
+			if step.Terminated.ExitCode == 0 {
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepSucceeded)
+			} else {
+				// failure
+				// (Terminated) emit&mark failed if exit code is not 0
+				ensureEventEmitted(ctx, sent, step.Name, resources.CloudEventTypeStepFailed)
+			}
+
+		}
+	}
+
+	// TODO proper reconcileErr handling
+	patch, err := getPatch(sent)
+	if err != nil {
+		return err
+	}
+	logger.Infof("PATCH: %s", patch)
+	_, err = client.Patch(taskrun.Name, types.MergePatchType, patch)
+	if err != nil {
+		logger.Errorf("failed to PATCH taskrun: %v", err)
+	}
+	return nil
```

Comment:
This is updating the TaskRun object ? :thinking:

Reply:
Yes, it does. I know this increases reconciliation queue for this controller and tekton taskrun controller. Do you think would it be a problem?

Reply:
One of the alternative I have currently is to define new CRD to store its state in etcd.

Reply:
@tom24d this stores the state of the event sent right ?

Reply:
@vdemeester
Yes, to the annotation field of TaskRun. The type of sent event is stored in the field to avoid duplicate of CloudEvent emission. Something like this:
```
annotation: &{[{step1 [dev.tekton.events.plugin.step.started.v1]} {step2 []} {step3 []}]}"
```
This is the snapshot after `step1` step gets started.

Reply:
right, I think it make sense then. We need to check how it affects the pipeline reconciliation loop (and impact on perf), but now I get it :stuck_out_tongue_closed_eyes:

Reply:
@vdemeester
OK, maybe I need more comment on the source code :) 
I will check the perf thing, then I move forward.
Thanks so much for taking your time!!